flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -885,6 +850,53 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
         }
     }
 
+    /**
+     * Transition into our final state in case of failure. We should be either in DEPLOYING,
+     * INITIALIZING, RUNNING, CANCELING, or FAILED loop for multiple retries during concurrent state
+     * changes via calls to cancel() or to failExternally()
+     */
+    private void transitionStateOnFailure(
+            Throwable t, AutoCloseableRegistry postFailureCleanUpRegistry) throws IOException {
+        while (true) {
+            ExecutionState current = this.executionState;
```

Comment:
I suggest adding a `yield()` call in the loop to prevent this thread hogging the cpu in a tight loop until done.